src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -423,7 +423,10 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
       }
     }
 
-    final def applyPhase(unit: CompilationUnit) = withCurrentUnit(unit)(apply(unit))
+    final def applyPhase(unit: CompilationUnit) = {
+      if (Thread.interrupted()) throw new InterruptedException
```

Comment:
Is it possible that instead of checking `Thread.interrupted` we check a variable in `Global`? Sometimes it's not possible to interrupt execution thread pools, and it's actually a costly operation to shut it down completely and then start it up again for another compilation run.

Reply:
If we do this, this would be highly beneficial for tools that control zinc like bloop or pants (/cc @stuhood) because we'd be able to set the variable in `Global` whenever we detect a cancellation in our task engine (in bloop's case, that's Monix).

Reply:
The surrounding tool need not shut down the threadpool.

```
var compilerThread: Thread = null
var interruped = false
threadPool.submit( () => {
  try {
    compilerThread = Thread.currentThread
    new Global().compile()
  }  catch { 
    case _: InterruptedException =>
      interrupted = false
  }
})

// elsewhere
Option(compilerThread).foreach(_.interrupt)
```

The calls in this PR to `Thread.interrupted` will clear the interrupted bit of the thread, so it is fine to continue work in the thread pool.

Reply:
Indeed, but the issue here is that the surrounding tool may not have access to the thread running the compiler. In the case of bloop, we use Monix and we cannot do `interrupt`. If anoter tool was using `Future`, it would have the same problem.

Reply:
You can capture the current thread just before you call `compile`, as I did above.

I suppose that, as a convenience, we could have `Run` capture the current thread at the start of in `compileUnitsInternal`, and then have a `Run.cancel()` method that interrupted that thread.

Reply:
> You can capture the current thread just before you call compile, as I did above.

Reply:
@retronym ping ^

Reply:
The latest implementation checks for both `Reporter.cancelled` and `Thread.interrupted`, so the caller can use whichever is more suitable.

That said, In the near term, it is unlikely that the compiler would transfer the main phase application loop to a different thread from the one that calls `compile`.